ocaml/typing/typecore.ml

```diff
@@ -5849,10 +5816,9 @@ and type_expect_
         exp_type = instance Predef.type_unit;
         exp_attributes = sexp.pexp_attributes;
         exp_env = env }
-  | Pexp_constraint (sarg, sty) ->
-      let type_mode, sty = alloc_mode_from_pexp_constraint_typ_attrs sty in
+  | Pexp_constraint (sarg, sty, modes) ->
       let (ty, exp_extra) =
-        type_constraint env sty type_mode
+        type_constraint env sty (Typemode.transl_alloc_mode modes)
```

Comment:
We need to constrain `sarg` to `modes`. The `modes` passed to `type_constraint` is only used for interpreting `sty` and nothing else. See the `type_mode_expr` function for reference.

Reply:
an issue im having when trying to do this is that `type_mode_expr` and `type_argument` are both `Parsetree.expression -> Typedtree.expression`, so I'm not really sure how to pass sarg through both

Reply:
Yeah `type_mode_expr` shouldn't call `type_expect`; instead, it should do some mode-related stuff and return a new `expected_mode`. This is similar to `let (ty, exp_extra) = type_constraint env sty type_mode`. Note this `ty` is later passed to `type_argument`. Similarly, we need a new `expected_mode` to be passed to `type_argument.`